src/raftstore/store/peer.rs

```diff
@@ -1028,22 +1019,16 @@ impl Peer {
                                 index));
         }
 
-        let engine = self.engine.clone();
-        let mut ctx = ExecContext {
-            snap: Snapshot::new(engine),
-            apply_state: self.get_store().apply_state.clone(),
-            wb: WriteBatch::new(),
-            req: req,
-            index: index,
-            term: term,
-        };
+        let mut ctx = ExecContext::new(self, index, term, req);
         let (mut resp, exec_result) = self.exec_raft_cmd(&mut ctx).unwrap_or_else(|e| {
             error!("{} execute raft command err: {:?}", self.tag, e);
             (cmd_resp::new_error(e), None)
         });
 
         ctx.apply_state.set_applied_index(index);
-        ctx.save(self.region_id).expect("save state must not fail");
+        if !self.pending_remove {
```

Comment:
If pending_remove, all the meta will be cleared, unnecessary to update the apply state. And we assume all the meta will be cleared when applying removal, updating apply state here will break this assumption.